codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamRingBuffer.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.handler.codec.http2;
+
+import io.netty.util.internal.MathUtil;
+
+import java.util.Arrays;
+
+/**
+ * Fixed size ring-buffer for {@code HTTP2} streams.
+ */
+final class Http2StreamRingBuffer {
+
+    private final int[] elements;
+    private int oldestIdx;
+
+    Http2StreamRingBuffer() {
+        this(32);
+    }
+
+    Http2StreamRingBuffer(int capacity) {
+        elements = new int[MathUtil.findNextPositivePowerOfTwo(capacity)];
+        Arrays.fill(elements, -1);
+    }
+
+    int capacity() {
+        return elements.length;
+    }
+
+    /**
+     * Add the given {@code streamId}
+     * @param streamId the ID
+     * @return {@code true} if could be added, false if it was not added as this ring-buffer already contained it.
+     */
+    boolean add(int streamId) {
+        assert streamId >= 0;
```

Comment:
Maybe this should be `> 0`? It seems unlikely we'll close the connection stream and still worry about this.